drivers/interrupt_controller/intc_mchp_ecia_xec.c

```diff
@@ -0,0 +1,1135 @@
+			girq->EN_CLR = BIT(bitpos);
+		}
+
+		/* clear GIRQ latched status */
+		girq->SRC = BIT(bitpos);
+		result &= ~BIT(bitpos);
+	}
+}
+#endif
+
+/**
+ * @brief initialize XEC ECIA driver
+ * NOTE: GIRQ22 is special used for waking the PLL from deep sleep when a
+ * peripheral receives data from an external entity (eSPI, I2C, etc). Once
+ * the data transfer is complete the system re-enters deep sleep unless the
+ * peripheral was configured to wake CPU after reception of data or event.
+ * GIRQ22 aggregated output and sources are not connected to the NVIC.
+ * We enable GIRQ22 aggregated output to ensure clock asynchronous wake
+ * functionality is operational.
+ */
+static int xec_ecia_init(const struct device *dev)
+{
+	const struct xec_ecia_config *cfg =
+		(const struct xec_ecia_config *const) (dev->config);
+	const struct device *const clk_dev = DEVICE_DT_GET(DT_NODELABEL(pcr));
+	struct ecs_regs *const ecs = ECS_XEC_REG_BASE;
+	struct ecia_regs *const ecia = (struct ecia_regs *)cfg->ecia_base;
+	uint32_t n = 0, nr = 0;
+	int ret;
+
+	ret = clock_control_on(clk_dev,
+			       (clock_control_subsys_t *)&cfg->clk_ctrl);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Enable all direct NVIC connections */
+	ecs->INTR_CTRL |= BIT(0);
+
+	/* gate off all aggregated outputs */
+	ecia->BLK_EN_CLR = UINT32_MAX;
+
+	/* connect aggregated only GIRQs to NVIC */
+	ecia->BLK_EN_SET = MCHP_ECIA_AGGR_BITMAP;
+
+	/* Clear all GIRQn source enables */
+	for (n = 0; n < MCHP_GIRQS; n++) {
+		ecia->GIRQ[n].EN_CLR = UINT32_MAX;
+	}
+
+	/* Clear all external NVIC enables and pending status */
+	nr = SCnSCB->ICTR;
+	for (n = 0u; n <= nr; n++) {
+		NVIC->ICER[n] = UINT32_MAX;
+		NVIC->ICPR[n] = UINT32_MAX;
+	}
+
+	return 0;
+}
+
+#define XEC_GIRQ_SOURCES(node_id, prop, idx)				\
+	(DT_PROP_BY_IDX(node_id, prop, idx) | BIT(7)),
+
+/* this one produces what we want */
+#define XEC_GIRQ_SOURCES2(node_id, prop, idx)				\
+	.sources[DT_PROP_BY_IDX(node_id, prop, idx)] =			\
+		((uint8_t)(idx) | BIT(7)),
+
+
+#if GIRQ08_AGGR
+#define GIRQ8_NSRCS DT_PROP_LEN(DT_NODELABEL(girq8), sources)
+
+static struct xec_girq_src_data xec_data_girq_8[GIRQ8_NSRCS];
+
+static const struct xec_girq_config xec_config_girq_8 = {
+	.base = DT_REG_ADDR(DT_NODELABEL(girq8)),
+	.girq_id = DT_PROP(DT_NODELABEL(girq8), girq_id),
+	.num_srcs = GIRQ8_NSRCS,
+	DT_FOREACH_PROP_ELEM(DT_NODELABEL(girq8), sources, XEC_GIRQ_SOURCES2)
+};
+
+static int xec_girq8_init(const struct device *dev)
+{
+	mchp_xec_ecia_girq_aggr_en(
+		GIRQ_ID_TO_BITPOS(DT_PROP(DT_NODELABEL(girq8), girq_id)), 1);
+
+	IRQ_CONNECT(DT_IRQN(DT_NODELABEL(girq8)),
+		    DT_IRQ(DT_NODELABEL(girq8), priority),
+		    xec_girq_isr,
+		    DEVICE_DT_GET(DT_NODELABEL(girq8)), 0);
+
+	irq_enable(DT_IRQN(DT_NODELABEL(girq8)));
+
+	return 0;
+}
+
+DEVICE_DT_DEFINE(DT_NODELABEL(girq8), xec_girq8_init,
+		 NULL, &xec_data_girq_8, &xec_config_girq_8,
+		 PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		 NULL);
+#endif
+#if GIRQ09_AGGR
```

Comment:
it seems this is severely repeated code

can these be consolidated to not do this?

Reply:
Maybe.  Each GIRQ had a different sized runtime data array.

Reply:
Found a way to consolidate with DT_FOREACH